libfabric can now find external installation

bvanessen opened the pull request:
Added support for libfabric to find an external installation and identify variants supported.

bvanessen committed:
Added support for libfabric to find an external installation and
identify variants supported.

becker33 commented:
@spackbot fix style

bvanessen committed:
[@spackbot] updating style on behalf of bvanessen

rajachan commented on var/spack/repos/builtin/packages/libfabric/package.py:
```diff
@@ -44,6 +48,7 @@ class Libfabric(AutotoolsPackage):
     version("1.4.2", sha256="5d027d7e4e34cb62508803e51d6bd2f477932ad68948996429df2bfff37ca2a5")
 
     fabrics = (
+        "cxi",
```
FWIW, I'm not in favor of adding this as a fabric option here as the cxi provider is still not available in upstream libfabric, where we pull the releases from. See discussion in https://github.com/spack/spack/pull/28446

I am not a spack expert, but this could be handled differently with a site-specific config and some combination of dynamically loading the external package and setting FI_PROVIDER=cxi at runtime??

bvanessen replied:
@rajachan I understand that the `cxi` code is not available upstream, I believe that we have had spack packages have variants and features for non-publicly accessible code bases.  I think that `mvapich2-gdr` and `cray-libsci` are an examples of such packages.  The key thing that I think that we would want to recognize is that CXI is a valid variant for some systems.  @becker33 @tgamblin How should we proceed?

rajachan replied:
Libfabric is not in the same bucket as other wholly binary/closed-source packages (and I am definitely not suggesting spack should not pull in binary/closed-source packages at all, that would be a silly stance to take :) ).  Adding a native 'fabrics' option for a provider that is not in the release tarball we are fetching does not seem right to me. 

I know I am being pedantic, and I really want to get this package to be as useful to folks as possible, so if this is the best way to handle it with spack and your use case then let's merge it!

rajachan replied:
cc: @shefty for visibility

shefty replied:
I'm unfamiliar with spack and the intent here.  But libfabric can make use of external provider libraries.  If there's some way to install a cxi package separately, and all this is doing is letting libfabric know that it might exists, this makes sense.  If this is assuming cxi is part of the libfabric package, then I agree adding it here may not make sense.  (Basically, I don't know. :)

rajachan replied:
A 'fabric' here is the equivalent of toggling `--enable-<provider>` so this is a build time thing. I also think this should use the provider as if it was an external provider, discovering at runtime.

bvanessen replied:
@rajachan and @becker33  Okay, I understand the issue here.  I have tried to mitigate it by making the CXI fabric only available when on a cray based system that has a slingshot fabric.  This still doesn't address the issue that the code isn't available for public builds, but should avoid impacting other instances of the package.  Thoughts?  Also it needs to be rebased on #33793 and #33939.

rajachan replied:
Yes, this looks a lot better! 👍

bvanessen commented:
@harisubramoni Here is the PR that I was mentioning.

harisubramoni commented:
Got it!

On Mon, Nov 14, 2022, 11:03 AM Brian Van Essen ***@***.***>
wrote:

> @harisubramoni <https://github.com/harisubramoni> Here is the PR that I
> was mentioning.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/spack/spack/pull/33796#issuecomment-1314084725>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEALULWA2F77Q6RB7EKQHMLWIJWFNANCNFSM6AAAAAAR33GR54>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

bvanessen committed:
Change the fabrics definition to only include CXI when on a cray
system with a libfabric-based slingshot network.

bvanessen commented:
@spackbot fix style

bvanessen committed:
[@spackbot] updating style on behalf of bvanessen

bvanessen commented:
@spackbot run pipeline

bvanessen commented:
@spackbot run pipeline

bvanessen committed:
Added a conflict when trying to build the CXI fabric value since it is
only available as closed source at this time.

Fixed a bug in how the get_options_from_variants was handling
conditional variants.  Now the conditional value is evaluated locally
to see if it should be a checked option.

bvanessen commented:
Rebased on #33939

bvanessen committed:
Added better handling of conditional variant options for the fabrics
variant.

bvanessen commented:
@spackbot fix style

bvanessen committed:
Fixed Flake8

bvanessen committed:
Handled the fabric variant options in a more careful manner to avoid
mismanaging the conditional options.

bvanessen committed:
Making the check for a conditional value more robust

bvanessen committed:
Fixed how the conditional value is handled

bvanessen commented:
@spackbot run pipeline

bvanessen commented:
@becker33 @haampie Can this PR get merged?

becker33 commented on var/spack/repos/builtin/packages/libfabric/package.py:
```diff
@@ -126,13 +160,42 @@ def configure_args(self):
             args.append("--with-kdreg=no")
 
         for fabric in self.fabrics:
```
What about `for fabric in [f if isinstance(f, str) else f[0].value for f in self.fabrics]:`

And then you can jump straight to the `if "fabrics=" + fabric in self.spec:` check with a lot less fuss.

bvanessen replied:
@becker33 Thanks for the suggestion.  I have updated it.

bvanessen committed:
Cleaned up with @becker33 commments
